src/lexer/disambiguateTokens.ts

```diff
@@ -35,7 +35,40 @@ const funcNameToKeyword = (token: Token, i: number, tokens: Token[]): Token => {
   if (token.type === TokenType.RESERVED_FUNCTION_NAME) {
     const nextToken = nextNonCommentToken(tokens, i);
     if (!nextToken || !isOpenParen(nextToken)) {
-      return { ...token, type: TokenType.RESERVED_KEYWORD };
+      return {
+        ...token,
+        type:
+          // Function names which are also data types
+          [
+            'BIGINT',
+            'BINARY',
+            'BIT',
+            'BLOB',
+            'BOOLEAN',
+            'CHAR',
+            'DATE',
+            'DECIMAL',
+            'DOUBLE',
+            'FLOAT',
+            'INT',
+            'INTEGER',
+            'JSON',
+            'NCHAR',
+            'NUMBER',
+            'NUMERIC',
+            'NVARCHAR',
+            'REAL',
+            'SMALLINT',
+            'TEXT',
+            'TIME',
+            'TIMESTAMP',
+            'TINYINT',
+            'VARCHAR',
+            'XML',
+          ].includes(token.text)
+            ? TokenType.RESERVED_DATA_TYPE
+            : TokenType.RESERVED_KEYWORD,
+      };
```

Comment:
Yeah, this is definitely problematic.

The root problem seems to be that these data types are also defined as functions, which previously was only done because we formatted those data-types like functions e.g. `DECIMAL(5, 2)`. Now that we have data types defined separately, there's really no reason to include them among the functions any more.

Reply:
another alternative approach that I thought of would be:

1. map each dialect to an array
   1. import + deduplicate all function names
   2. import + deduplicate all data type names
   3. find the overlapping entries
2. merge all of the arrays and deduplicate

this would make the items more accurate

but I think you're suggesting to go in a different direction with the grammar... I don't think I fully understand what you mean yet

Reply:
Yes, I would indeed suggest this last option:

> Y) another way of interpreting your comment leads me to believe you're suggesting removing `DECIMAL` (and anything else that is a data type) from `*.functions.ts` - but I think that also would be problematic...?

Simply removing them will of course result in `DECIMAL(1, 2)` being formatted as `DECIMAL (1, 2)` as the formatter no more recognizes as formats it as a function. What should be done, is also changing the parser so it recognizes both `<func-name> <parenthesis>` and `<data-type> <parenthesis>` as function calls. Or alternatively we could also introduce a separate node like `ParameterizedDataType` for the latter.

I'm happy to look into this by myself as I'm now finally back at my computer, and I have a hunch that there are some quirks of the parser that will also be getting in the way of this.

Reply:
Ok thanks!

One thing to keep in mind and decide on is what the behavior of conflicting options would be in that `DECIMAL(1, 2)` case, eg:

```js
{
  functionCase: 'lower',
  dataTypeCase: 'upper',
}
```

Example query

```sql
SELECT DeCimal(1, 2) AS dec, noW() AS n
```

I'm guessing it would be this:

```sql
SELECT DECIMAL(1, 2) AS dec, now() AS n
```

Reply:
Yep. Definitely like so. Treating `DECIMAL()` as a function is an implementation detail of the formatter. It would very much confuse users when `dataTypeCase` doesn't affect parameterized data types.